### PR TITLE
Update ICICLE integration to use v3 ICICLE

### LINKED ISSUE #1166: bug: Icicle integration is out of date
The Icicle integration for the Groth16/BN254 backend is out of date and does not currently compile, and the version of icicle that gnark depends on is old.

## Description

Trying to build gnark with `-tags=icicle` fails due to accumulated changes in the rest of the gnark internal API that were not reflected in the Icicle backend. See the specific compile errors mentioned in "Actual Behavior" below.

Additionally, the icicle dependency used by gnark is currently commit https://github.com/ingonyama-zk/icicle/commit/97f0079e5c71d0bd4b8b97fababcfdb9940f12e1 (specified in [go.mod here](https://github.com/Consensys/gnark/blob/db299cef6c78dc5acff8453b66c910c15ea88123/go.mod#L29), indirectly from the [iciclegnark](https://github.com/ingonyama-zk/iciclegnark) dependency), which means that the current instructions specified [here](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/golang) do not work for gnark: the current version of the icicle Go bindings create static libraries like `libingo_curve_bn254.a` and `libingo_field_bn254.a` while the older version that gnark relies on creates a shared library `libbn254.so` and the build process is documented [here](https://github.com/ingonyama-zk/icicle/blob/97f0079e5c71d0bd4b8b97fababcfdb9940f12e1/goicicle/README.md) instead. It would be very nice if the gnark integration could be updated to support the latest icicle version as well, besides simply fixing the compile errors mentioned in this issue, though that probably requires Ingonyama to make the respective updates to the iciclegnark package.

## Expected Behavior
The code should compile without errors and generated proofs should verify correctly.

## Actual Behavior
Building with `-tags=icicle` gives the following compile errors:
```
# github.com/consensys/gnark/backend/groth16/bn254/icicle [github.com/consensys/gnark.test]
backend/groth16/bn254/icicle/icicle.go:44:30: cannot use pk.Domain.CosetTableInv (value of type func() ([]"github.com/consensys/gnark-crypto/ecc/bn254/fr".Element, error)) as []"github.com/consensys/gnark-crypto/ecc/bn254/fr".Element value in argument to iciclegnark.CopyToDevice
backend/groth16/bn254/icicle/icicle.go:47:30: cannot use pk.Domain.CosetTable (value of type func() ([]"github.com/consensys/gnark-crypto/ecc/bn254/fr".Element, error)) as []"github.com/consensys/gnark-crypto/ecc/bn254/fr".Element value in argument to iciclegnark.CopyToDevice
backend/groth16/bn254/icicle/icicle.go:160:73: commitmentInfo[i].HintID undefined (type constraint.Groth16Commitment has no field or method HintID)
```

## Possible Fix

I tried fixing this locally in this commit https://github.com/wwared/gnark/commit/186e9de8321f7933e42f4d1b01c40a25f33fd4a5

However, even though this enabled me to compile gnark with Icicle support (linking against the older version at commit https://github.com/ingonyama-zk/icicle/commit/97f0079e5c71d0bd4b8b97fababcfdb9940f12e1), proofs generated with this would fail to verify with `panic: pairing doesn't match`, so this attempted fix is either wrong or incomplete.

## Steps to Reproduce

1. Run `go test -tags=icicle` on the repository root of current master, or attempt to build a gnark-using program by passing `-tags=icicle` to the build/run command line.

## Context
I am trying to generate a proof using GPU acceleration for a circuit defined on the latest gnark. It's not easy for me to revert to an older gnark version.

## Your Environment
* gnark version used: `HEAD@master` (https://github.com/Consensys/gnark/commit/db299cef6c78dc5acff8453b66c910c15ea88123 as of writing)
* gnark-crypto version used: Whichever one is used by default (`v0.12.2-0.20240504013751-564b6f724c3b` according to `go.mod`)
* go version: 1.22.3
* Operating System and version: Linux 6.6.30

### PR DESCRIPTION
**Pull request by jeremyfelder**
# Description

This PR updates the ICICLE integration to use v3 of ICICLE.

Fixes #1166

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

# How has this been tested?

```go
package main

import (
	"fmt"
	"time"

	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/backend"
	"github.com/consensys/gnark/backend/groth16"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/frontend/cs/r1cs"
)

// CubicCircuit defines a simple circuit
// x**3 + x + 5 == y
type CubicCircuit struct {
	// struct tags on a variable is optional
	// default uses variable name and secret visibility.
	X frontend.Variable `gnark:"x"`
	Y frontend.Variable `gnark:",public"`
}

// Define declares the circuit constraints
// x**3 + x + 5 == y
func (circuit *CubicCircuit) Define(api frontend.API) error {
	x3 := api.Mul(circuit.X, circuit.X, circuit.X)
	for i := 0; i < 1000000; i++ {
		api.AssertIsEqual(circuit.Y, api.Add(x3, circuit.X, 5))
	}
	return nil
}

func main() {
	// compiles our circuit into a R1CS
	var circuit CubicCircuit
	ccs, _ := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit)

	// groth16 zkSNARK: Setup
	pk, vk, _ := groth16.Setup(ccs)

	// witness definition
	assignment := CubicCircuit{X: 3, Y: 35}
	witness, _ := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
	publicWitness, _ := witness.Public()
	// groth16: Prove & Verify
	icicleTimeStart := time.Now()
	proof, err := groth16.Prove(ccs, pk, witness, backend.WithIcicleAcceleration())
	fmt.Println("Icicle proof time:", time.Since(icicleTimeStart))

	if err != nil {
		fmt.Println(err)
	}

	if err = groth16.Verify(proof, vk, publicWitness); err != nil {
		panic("Failed verification")
	}
	
	gnarkTimeStart := time.Now()
	proofNoIcicle, err := groth16.Prove(ccs, pk, witness)
	fmt.Println("Gnark CPU proof time:", time.Since(gnarkTimeStart))
	
	if err != nil {
		fmt.Println(err)
	}

	if err = groth16.Verify(proofNoIcicle, vk, publicWitness); err != nil {
		panic("Failed verification")
	}
}

```

# How has this been benchmarked?

- [x] Using the above example circuit on a i9-13900K, 64GB RAM + NVIDIA RTX 4080
image

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [ ] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by p4u** [comment]
hey @jeremyfelder, can you try with bls12377 ?

Because I managed to make it work on bn254, but the same code on bls12377, icicle acceleration is just ignored.

I'm compiling the Go icicle wrapper with `./build -curves=all` so it should be available (acually I see the compilation on the bls12377).

**Comment by jeremyfelder** [comment]
> hey @jeremyfelder, can you try with bls12377 ?
> 
> Because I managed to make it work on bn254, but the same code on bls12377, icicle acceleration is just ignored.
> 
> I'm compiling the Go icicle wrapper with `./build -curves=all` so it should be available (acually I see the compilation traces on the bls12377).

Great to hear that you got bn254 working 🔥!
Right now the gnark<>ICICLE integration only includes bn254. It should be relatively simple to add all the curves/fields that ICICLE supports though I didn't want to bloat this PR with something that should be designed better to stay updated as @ivokub and I have discussed in the past.

**Comment by ivokub** [comment]
> > hey @jeremyfelder, can you try with bls12377 ?
> > Because I managed to make it work on bn254, but the same code on bls12377, icicle acceleration is just ignored.
> > I'm compiling the Go icicle wrapper with `./build -curves=all` so it should be available (acually I see the compilation traces on the bls12377).
> 
> Great to hear that you got bn254 working 🔥! Right now the gnark<>ICICLE integration only includes bn254. It should be relatively simple to add all the curves/fields that ICICLE supports though I didn't want to bloat this PR with something that should be designed better to stay updated as @ivokub and I have discussed in the past.

Yep - I'll try to review and merge this PR promptly and then figure out a bit better to switch between accelerated and non-accelerated implementations. Or see if the current approach scales to different curves nicely.

File: backend/groth16/bn254/icicle/icicle.go [hunk 1]
```diff
@@ -202,33 +267,44 @@ func Prove(r1cs *cs.R1CS, pk *ProvingKey, fullWitness witness.Witness, opts ...b
 	wireValues := []fr.Element(solution.W)
 
 	start := time.Now()
+	poks := make([]curve.G1Affine, len(pk.CommitmentKeys))
 
+	for i := range pk.CommitmentKeys {
+		var err error
+		if poks[i], err = pk.CommitmentKeys[i].ProveKnowledge(privateCommittedValues[i]); err != nil {
```
Comment (ivokub) [inline backend/groth16/bn254/icicle/icicle.go:274]: In essence this is a big MSM, in this PR we still do it in CPU.

For big circuits which use field emulation this MSM can be huge (hundreds of thousands of inputs), so should be worthwile to move to ICICLE as well?

File: backend/groth16/bn254/icicle/icicle.go [hunk 2]
```diff
@@ -142,56 +207,56 @@ func Prove(r1cs *cs.R1CS, pk *ProvingKey, fullWitness witness.Witness, opts ...b
...
+		in = in[1:]
+		privateCommittedValues[i] = make([]fr.Element, len(commitmentInfo[i].PrivateCommitted))
+		hashed := in[:len(commitmentInfo[i].PublicAndCommitmentCommitted)]
+		committed := in[+len(hashed):]
+		for j, inJ := range committed {
+			privateCommittedValues[i][j].SetBigInt(inJ)
+		}
+
+		var err error
+		if proof.Commitments[i], err = pk.CommitmentKeys[i].Commit(privateCommittedValues[i]); err != nil {
```
Comment (ivokub) [inline backend/groth16/bn254/icicle/icicle.go:244]: This is also a MSM

**Pull request merged**